crates/oxc_parser/src/lexer/source.rs

```diff
@@ -499,6 +499,29 @@ impl<'a> Source<'a> {
         }
     }
 
+    /// Peek next two bytes of source without consuming them.
+    #[inline]
+    pub(super) fn peek_byte2(&self) -> Option<[u8; 2]> {
+        let will_reach_eof = {
+            if self.ptr as isize == isize::MAX || self.ptr == self.end {
+                true
+            } else {
+                // SAFETY: the check above ensured that the resulting pointer:
+                //  * does not overflow an `isize`, and
+                //  * is either in bounds or one byte past the end.
+                unsafe { self.ptr.add(1) == self.end }
+            }
+        };
+
+        if will_reach_eof {
+            None
+        } else {
+            // SAFETY: Safe to read from `ptr` as we just checked it's not out of bounds
+            let bytes = unsafe { self.position().read2() };
+            Some(bytes)
+        }
+    }
```

Comment:
Also, we should probably rename this function to `peek_2_bytes` because `peek_char2` is "peek the char after the current one" rather than "peek the next 2 chars", so it's a bit confusing if they're called similar things when what they do is different.

Reply:
Makes sense, I'll rework it that way.

You are right regarding the max-pointer check being buggy, I misread the safety requirements at https://doc.rust-lang.org/core/primitive.pointer.html#method.add (it's the intermediate offset that must not overflow `isize`, not the resulting pointer).